app/build.gradle

```diff
@@ -38,6 +38,9 @@ android {
             useSupportLibrary true
         }
     }
+    kotlinOptions {
+        freeCompilerArgs = ["-Xallow-result-return-type"]
```

Comment:
I wonder if we should be using the `Result` class directly, rather than another class? I ask because this seems to be required to work around [KEEP-127](https://github.com/Kotlin/KEEP/blob/master/proposals/stdlib/result.md#limitations):

> Result<T> cannot be used as a direct result type of Kotlin functions, properties of Result type are also restricted

Reply:
Correct, this is me being lazy, really 😝 
We can define another type but that ultimately would yield the same as `kotlin.Result`, the only usage of that is in the instantiation of the `CastContext`, that can fail (a similar concept in other languages exist and is better represented, for example the [`Try` construction in scala](https://www.scala-lang.org/api/2.9.3/scala/util/Try.html)).

The reason the Kotlin team requires the compiler flag to allow for usages in return types is because of the possible API-breakage, as the docs describe:

> The rationale behind these limitations is that future versions of Kotlin may expand and/or change semantics of functions that return Result type and null-safety operators may change their semantics when used on values of Result type. In order to avoid breaking existing code in the future releases of Kotlin and leave door open for those changes, the corresponding uses produce an error now. 

I can fairly easily replace that `kotlin.Result` with another construction. 
 🚧 🔨